src/metabase/search/postgres/index.clj

```diff
@@ -0,0 +1,124 @@
+(ns metabase.search.postgres.index
+  (:require
+   [clojure.string :as str]
+   [honey.sql :as sql]
+   [honey.sql.helpers :as sql.helpers]
+   [toucan2.core :as t2]))
+
+(def ^:private active-table :search_index)
+
+(def ^:private pending-table :search_index_next)
+
+(def ^:private retired-table :search_index_retired)
+
+(defonce ^:private initialized? (atom false))
+
+(defonce ^:private reindexing? (atom false))
+
+(def ^:private tsv-language "english")
+
+(defn- exists? [table-name]
+  (t2/exists? :information_schema.tables :table_name (name table-name)))
+
+(defn- drop-table! [table-name]
+  (boolean
+   (when (exists? table-name)
+     (t2/query (sql.helpers/drop-table table-name)))))
+
+(defn- rename-table! [old new]
+  (when (and (exists? old) (not (exists? new)))
+    (-> (sql.helpers/alter-table old)
+        (sql.helpers/rename-table new)
+        t2/query)))
+
+(defn create-pending!
+  "Create a non-active search index table."
+  []
+  (when (not @reindexing?)
+    (when-not (exists? pending-table)
+      (-> (sql.helpers/create-table pending-table)
+          (sql.helpers/with-columns
+            [[:id :bigint [:primary-key] [:raw "GENERATED BY DEFAULT AS IDENTITY"]]
+             ;; entity
+             [:model_id :int :not-null]
+             [:model [:varchar 254] :not-null] ;; TODO find the right size
+             ;; search
+             [:search_vector :tsvector :not-null]
+             ;; scoring related
+             [:model_rank :int :not-null]
+             ;; permission related entities
+             [:collection_id :int]
+             [:database_id :int]
+             [:table_id :int]
+             ;; filter related
+             [:archived :boolean]])
+          t2/query)
+
+      (t2/query
+       (format "CREATE INDEX IF NOT EXISTS %s_tsvector_idx ON %s USING gin (search_vector)"
+               (str/replace (str (name active-table) "_" (random-uuid)) #"-" "_")
+               (name pending-table))))
+    (reset! reindexing? true)))
+
+(defn activate-pending!
+  "Make the pending index active, if it exists. Returns true if it did so."
+  []
+  ;; ... just in case it wasn't cleaned up last time.
+  (drop-table! retired-table)
+  (when (exists? pending-table)
+    (t2/with-transaction [_conn]
+      (rename-table! active-table retired-table)
+      (rename-table! pending-table active-table))
+    (reset! reindexing? false)
```

Comment:
Maybe we should do this before the `with-transaction`?

Also, why `drop-table!` is not executed in the transaction?

Reply:
Hmm, I guess we could stop reindexing before. My thinking was that we don't know why the transaction failed so its safer no to do anything.

I haven't tested this for Postgres, but for some databases DROP is not transactional, and for others it can lock the table and cause concurrent queries to fail. This pattern is safe across all databases as far as I know, and simple enough.